### PR TITLE
fix(kg): give NetworkX's reload fence a channel the manager cannot lose

### PR DESCRIPTION
**Pull request by danielaskdd**
## Description

`NetworkXStorage`'s cross-process fence rested entirely on the `storage_updated` flag. `set_all_update_flags` publishes it with one Manager RPC per process, so it can fail — and fail **partway**, leaving some workers notified and the rest not. Both places that consult it read the same flag: `_get_graph`, the only point a process picks up a fresher snapshot, and `index_done_callback`'s decline branch.

An unnotified worker therefore did not reload, did not decline, and wrote the **whole** GraphML file — silently overwriting the durable commit it never saw.

*Single writer per workspace* does not prevent that: it gives one writer **at a time**, not always the same process, so a later request legitimately makes another worker the next writer. The same mechanism has a second, larger window that needs no failure at all — the flag is only evaluated when a commit happens to have flipped it, so a process that misses one notification stays stale with no upper bound, while the authoritative surface (the file) is never consulted.

This adds the file itself as a second channel: `(st_mtime_ns, st_size)` compared against the fingerprint recorded when this process last loaded or wrote it. The GraphML file is already documented as the only cross-process synchronization surface, and unlike the flag it is *state* rather than a consumable event, so a failed notification cannot disable it.

**Both channels stay, permanently.** Their blind spots do not overlap:

| | flag | fingerprint |
|---|---|---|
| Semantics | event: someone committed | state: the file is not the one I loaded |
| Can be lost | yes — this issue | no, while the file exists |
| Blind spot | transient manager outage | two commits inside one timestamp tick |

The flag fails on a manager outage, where the two commits are far apart in time and the fingerprint sees them; the fingerprint fails when two commits land in one tick, which needs a healthy, fast-committing system — exactly when the flag works. The flag is read first (accelerator), the fingerprint is the one that must never be skipped (authoritative).

## Related Issues

Phase 1 of #3854 (NetworkX). Phase 2 covers `NanoVectorDBStorage` / `FaissVectorDBStorage`, which are exposed in the same shape including on their `for_write=True` path.

Orthogonal to #3838 R3, which serializes admin-vs-admin writers: R3 removes the *concurrent-overlap* variant, while the *sequential-handoff* variant — two admin requests a second apart on different workers — survives it untouched and is the one this fixes. Recorded in both issues.

## Changes Made

**`lightrag/kg/shared_storage.py`** — new `is_multiprocess_mode()` accessor (additive; nothing else touched).

**`lightrag/kg/networkx_impl.py`**
- `_stat_fingerprint` / `_adopt_fingerprint` / `_peer_commit_detected` / `_reload_locked` / `_record_fingerprint`.
- Wired into `_get_graph`, `index_done_callback`'s decline branch, the commit hook, the failed-save recovery path, and `drop`.
- A per-process counter, logged at WARNING on every `flag == False && fingerprint changed` — so a deployment can tell whether this window actually occurs instead of only being reachable on paper.
- *Cross-process sync protocol* docstring rewritten for two channels, with both accepted residues and their recovery paths.

Four rules the implementation holds to, each with a test that goes red without it:

1. **One reload, one post-condition**, whichever channel fired: `_reload_locked` records the fingerprint **and** clears the flag. Open-coding either half costs a full GraphML re-parse on the next call.
2. **The fingerprint is sampled before the file is read, never after.** Sampled after the parse it can belong to a newer file than the one in memory, and recording it would suppress the reload that file needs — the one way this change could *introduce* a lost write.
3. **The writer adopts the fingerprint of the file it just wrote**, before the fallible notification, so it does not read its own commit as a peer's.
4. **Gated on `is_multiprocess_mode()`** — single-process mode has no peer that could have committed, and reloading for an externally edited file would discard this process's own uncommitted mutations.

`st_ino` is deliberately **not** part of the fingerprint: `atomic_write` renames a tmp file over the target, freeing the previous inode, and the allocator hands that same inode straight back to the next tmp file — measured at 28 reuses across 30 consecutive commits, so it discriminates nothing.

**`nano_vector_db_impl.py` / `faiss_impl.py`** — docstrings only, no behavior change: their protocol is now labelled flag-only with a pointer to phase 2. The old "identical in shape" cross-reference in `NetworkXStorage` would otherwise have gone stale.

**`tests/kg/networkx_impl/test_networkx_fingerprint_fence.py`** — 8 tests: the stale writer declining instead of overwriting, the reader recovering a commit it was never told about, the three rules above, and the two residues (tick collision, unreadable `stat`) plus single-process inertness.

## Behavior change to expect

The decline branch now refuses a stale writer instead of overwriting. That surfaces as an error through `_commit_graph_or_raise` where the write was previously lost in silence — the intended direction, and concentrated in the concurrent-overlap variant #3838 R3 removes. Turning a declined admin commit into "reload and retry once" belongs to R3's 409/retry-semantics discussion, not here.

## Checklist

- [x] Changes tested locally
- [x] Code reviewed
- [x] Documentation updated (class docstrings; #3854 carries the design record)
- [x] Unit tests added

## Additional Notes

**Every new test was proven able to fail.** Disabling `_peer_commit_detected` turned the 4 fence tests red (including the core overwrite one); removing the writer's `_record_fingerprint()` from the commit hook and from `drop` turned the other 2 red — that run also exposed that the `drop` test was vacuous, so it now asserts the fingerprint is non-`None` before the drop.

**Test results.** `tests/kg/networkx_impl` → **112 passed** (104 pre-existing + 8 new). `tests/pipeline` → **773 passed**. `ruff check lightrag/` clean.

**On the full suite in this sandbox:** because `shared_storage.py` is listed as cross-cutting in `AGENTS.md`, I ran `./scripts/test.sh tests` → 143 failed / 7928 passed / 199 skipped. All 143 are environmental, not regressions: reaching them required installing the `offline-storage` / `offline-llm` extras (62 modules could not even be collected without them), and those extras activate code paths that download tokenizers and spaCy models, which the sandbox proxy refuses — every one of the 143 traces to `OSError: Tunnel connection failed: 403 Forbidden`, and none mentions either changed module. They cluster in `tests/parser/docx/test_smart_heading_*`, `tests/llm/*tokenizer*`, `tests/llm/openai_impl/*`, `tests/extraction/*`, `tests/kg/test_batch_graph_operations.py` and `tests/kg/noop_impl/`. CI is the real arbiter for the full suite.

**Measurements behind the design** (dev container, Linux 6.18.44, `Manager` over a local socket) — full data in #3854:
- `os.stat` **0.8 µs** vs the flag read **66 µs** (Manager RPC) vs `_storage_lock` acquire+release **354 µs**. `_get_graph` already costs ~420 µs per call in multiprocess mode, so the added `stat` is +0.2%; the flag is the expensive channel, which is why cost is not an argument for treating it as primary.
- `st_mtime_ns` resolution is kernel-dependent across an order of magnitude (~10 µs on Linux ≥ 6.13 multigrain timestamps, 1–4 ms on older kernels, 1 s on ext3/HFS+, ~15.6 ms system-clock tick on Windows). Hence the tick-collision residue, and the mandatory collision test that forces it with `os.utime` rather than leaving it to wall-clock luck.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01PyHEH7RFCxPzm2bup76J4K

---
_Generated by [Claude Code](https://claude.ai/code/session_01PyHEH7RFCxPzm2bup76J4K)_

### CODE REVIEW

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/kg/networkx_impl.py [hunk 1]
```diff
@@ -935,30 +1154,52 @@ async def index_done_callback(self) -> bool:
...
+                logger.warning(
+                    f"[{self.workspace}] Declining to save graph "
+                    f"{self._graphml_xml_file}: the file on disk is not the one "
+                    "this process loaded and no reload notification arrived for "
+                    "it, so saving would overwrite another process's commit. "
+                    "Reloading and reporting the commit as declined (occurrence "
+                    f"#{self._missed_notification_reloads} in this process)."
+                )
+                self._reload_locked()
+                return False
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/kg/networkx_impl.py:1211]: Correct, and it mattered more than the finding states — thanks. Fixed in 06b301e.

Verified: `_flush_one` (`lightrag/lightrag.py:3439`) awaited `index_done_callback()` and discarded the return, so only exceptions were detectable. Only `NetworkXStorage` ever returns `False`; every other backend returns `None` or `True`, and `BaseGraphStorage.index_done_callback` is declared `-> None`.

The part worth naming explicitly: before this PR the decline branch was unreachable in a writer (it needed a flag that is permanently `False` there under the single-writer contract), so the gap was latent. The fingerprint channel makes it live — and shipping it without this fix would not have narrowed the residue, it would have **swapped it for its mirror**:

- before: the stale writer overwrites → the peer's commit is lost, this document's writes are genuinely durable;
- fence alone: the peer's commit is preserved → this document's graph writes are discarded and it is still marked PROCESSED, with nothing to recover them from.

`AGENTS.md`'s *Consistency without transactions* rules that out — "a change must improve on an accepted residue, not swap it for its mirror". With the decline propagated the residue heals instead: the batch takes the FAILED path and its reprocessing re-extracts and re-writes the discarded work.

Fixed in `_flush_one` rather than by raising inside the backend, for the reason your finding suggests as the alternative: the defect is a caller ignoring a documented return value, so fixing it there covers every flush path and every future backend, and matches the stance `utils_graph._commit_graph_or_raise` already takes on the admin paths. Identity, not truthiness (`committed is False`) — testing `None` for falsiness would fail every flush in the project.

Also recorded in `NetworkXStorage`'s *Cross-process sync protocol* docstring, so the requirement carries into phase 2 (Nano/FAISS decline paths must propagate the same way), plus tests: `test_insert_done_rejects_a_declined_commit`, and `test_insert_done_accepts_none_and_true` pinning the identity rule. Both were confirmed to go red without the fix.


---
_Generated by [Claude Code](https://claude.ai/code)_

File: lightrag/kg/networkx_impl.py [hunk 2]
```diff
@@ -326,40 +424,161 @@ def _gate(self) -> asyncio.Event:
...
+            # report a completed destruction as an error -- a broken log sink
+            # included.
+            log_without_raising(
+                logger.warning,
+                f"[{self.workspace}] Could not stat {self._graphml_xml_file} "
+                "to check for peer commits; the reload fence falls back to the "
+                f"notification flag alone: {exc}",
+            )
+            return _FINGERPRINT_UNREADABLE
+        return (st.st_mtime_ns, st.st_size)
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/kg/networkx_impl.py:469]: Declining this one — the collision is a deliberately accepted residue, not an oversight. Maintainer decision; recording the reasoning so it is not re-derived.

**Content identity is too expensive to be the fence.** The fence is evaluated inside `_get_graph`, which every read and every write funnels through — thousands of times in one query. A content hash means reading and digesting the whole GraphML file per graph operation, against `0.8 µs` for the `stat` it replaces. That is not a trade this fence can make; the cost is what keeps it usable at all.

**The fingerprint is the fallback channel, and a missed detection is acceptable by design.** The flag stays the primary notification; the file check exists to catch the case where the flag was lost. A fence that only *usually* catches that is still strictly better than one that never does.

**The residue is a strict subset of the window that exists today, not a new hazard.** Before this PR, a lost notification *alone* lost the write. After it, losing the write additionally requires the two commits to land inside one filesystem timestamp tick **and** to produce an identical file size. Every case outside that intersection is now caught. And in the collision case the outcome is exactly the pre-PR behaviour — never worse than what is already shipped.

Note also that with 06b301e (your other finding) the collision case is the only remaining silent one: everywhere the fence *does* fire, the declined commit is now propagated as a failure and the work is reprocessed.

**On the generation marker specifically** — it is not dismissed, it is recorded as the option in #3854 with its cost: the writer can `stat` before the write and, after `os.replace`, bump with `os.utime` if the new mtime did not advance, which makes mtime a strictly monotonic version number regardless of clock granularity. Cheap (one extra `stat`, a rare `utime`). Not taken here because it writes synthetic metadata, and on 1 s / 2 s-granularity filesystems (ext3, HFS+, FAT) a +1 ns bump is truncated so a whole-granule bump would be needed, producing user-visible future timestamps. If the deployment counter added in this PR (`flag == False && fingerprint changed`, logged at WARNING) ever shows this window occurring in practice, that option is where to go next.

The residue and its recovery path are written into `NetworkXStorage`'s *Cross-process sync protocol* docstring, per `AGENTS.md` (a documented residue is a decision, an undocumented one is a defect), and pinned by a test that forces the collision with `os.utime` rather than leaving it to wall-clock luck.


---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by danielaskdd** [comment]
@codex review

File: lightrag/kg/networkx_impl.py [hunk 3]
```diff
@@ -1069,6 +1322,11 @@ async def _committed() -> None:
                     # could mistake the unpersisted mutation for durable state
                     # and sweep the live object's tracking row.
                     self.storage_updated.value = True
+                    # Re-arm the file channel as well, so the fence holds even
+                    # if the flag assignment above failed too (both go through
+                    # the same manager). None differs from any real file, so
+                    # the next _get_graph retries the reload on its own.
+                    self._loaded_fingerprint = None
```
[the remark this answers is not in the document]
Comment (danielaskdd) [reply lightrag/kg/networkx_impl.py:1339]: Valid — fixed in 89be499. This one is a self-inconsistency in the change: the same "local write first, fallible RPC second" rule is applied and documented in the commit hook (`_record_fingerprint()` before `set_all_update_flags`), and then not applied here. The comment on those lines even claimed the property the ordering defeats — *"so the fence holds even if the flag assignment above failed too"* — which it cannot, since the raising statement came first.

Confirmed reachable as you describe, with one refinement worth recording: a manager that is already gone makes `index_done_callback` fail at its very first flag *read*, before the save. The sequence that gets here is the manager dying **during** the offloaded save — the flag read succeeded, `write_nx_graph` failed, the recovery reload failed, and the flag write is now the first RPC to notice. The regression test models exactly that (reads succeed, writes raise) rather than a proxy that is dead throughout, which would never reach the save.

Changes:

- `self._loaded_fingerprint = None` now precedes the flag write. It is a plain attribute assignment that cannot fail with the manager, so the file channel is armed unconditionally and the next `_get_graph` reloads instead of trusting the divergent graph.
- The flag write is now wrapped. It can only fail in multiprocess mode — single-process flags are a local `MutableBoolean` — which is exactly where the file channel is armed and covers it. Unguarded it also **masked the save error**: the `BrokenPipeError` propagated out of the `except` block in place of the `OSError` the caller must see, contradicting the "Report, never mask" note two lines above. It is now logged through `log_without_raising` and the save error reaches the caller.

Test: `test_failed_save_arms_the_file_channel_even_if_the_flag_write_fails` asserts the flag arming was attempted and failed, that `_loaded_fingerprint` is armed regardless, and that `OSError("save boom")` is what surfaces. Confirmed to go red with the old ordering.

`tests/pipeline` + `tests/kg/networkx_impl` → 889 passed.


---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by danielaskdd** [comment]
@codex review
